src/evm_parse.c

```diff
@@ -418,6 +475,223 @@ enum parse_rv parse_rlp_txn(struct EVM_RLP_list_state *const state, evm_parser_m
+
+            // If we exhaust the apdu while parsing the length, there's nothing yet to hand to the subparser
+            if(state->rlpItem_state.state < 2)
+              return sub_rv;
+
+            if(state->hasTo) {
+              if(meta->known_destination) {
+                if(state->rlpItem_state.do_init && meta->known_destination->init_data)
+                  ((known_destination_init)PIC(meta->known_destination->init_data))(&(state->rlpItem_state.endpoint_state), state->rlpItem_state.length);
+                PRINTF("INIT: %u\n", state->rlpItem_state.do_init);
+                PRINTF("Chunk: [%u] %.*h\n", state->rlpItem_state.chunk.length, state->rlpItem_state.chunk.length, state->rlpItem_state.chunk.src);
+                if(meta->known_destination->handle_data) {
+                  PRINTF("HANDLING DATA\n");
+                  sub_rv = ((known_destination_parser)PIC(meta->known_destination->handle_data))(&(state->rlpItem_state.endpoint_state), &(state->rlpItem_state.chunk), meta);
+                }
+                PRINTF("PARSER CALLED [sub_rv: %u]\n", sub_rv);
+              }
+              else {
+                struct EVM_RLP_item_state *const item_state = &state->rlpItem_state;
+                struct EVM_ABI_state *const abi_state = &item_state->endpoint_state.abi_state;
+                if(item_state->do_init)
+                  init_abi_call_data(abi_state, item_state->length);
+
+                if(abi_state->data_length == 0) {
+                  sub_rv = PARSE_RV_DONE;
+                  state->item_index++;
+                  init_rlp_item(&state->rlpItem_state);
+                  ADD_ACCUM_PROMPT("Transfer", output_evm_prompt_to_string);
+                  return PARSE_RV_PROMPT;
+                }
+                else {
+                  sub_rv = parse_abi_call_data(abi_state,
+                                               &item_state->chunk,
+                                               meta,
+                                               !zero256(&state->value));
+                }
+              }
+            }
+
+            // Can't use the FINISH_ITEM_ macro here because we need to do a prompt in the middle of it.
+            if(sub_rv != PARSE_RV_DONE) return sub_rv;
+            state->item_index++;
+            uint64_t len = state->rlpItem_state.length;
+            if(!state->hasTo) {
+              SET_PROMPT_VALUE(entry->data.output_prompt.calldata_preview.cropped = len > MAX_CALLDATA_PREVIEW);
+              SET_PROMPT_VALUE(entry->data.output_prompt.calldata_preview.count = MIN(len, (uint64_t)MAX_CALLDATA_PREVIEW));
+              SET_PROMPT_VALUE(memcpy(entry->data.output_prompt.calldata_preview.buffer,
+                                      &state->rlpItem_state.buffer,
+                                      entry->data.output_prompt.calldata_preview.count));
+            }
+            state->hasData = len > 0;
+            init_rlp_item(&state->rlpItem_state);
```

Comment:
Might be able to factor this into a standalone function for sharing with the legacy tx parser. 
I _think_ you only need to ensure `meta` has the same name inside that function so that `SET_PROMPT_VALUE` will find it, but if it's a lot more work it's probably not worth it right now. 
Applies to half this parser really.

Reply:
Added a couple refactors - some stuff that I initially thought I could extract, I ended up being unable to refactor easily, due to embedded `return` statements for example. I'll spend a little time looking for more spots to extract.